logstash-rs/src/output/tcp.rs

```diff
@@ -0,0 +1,180 @@
+use crate::prelude::*;
+use std::fmt::Write as FMTWrite;
+use std::io::Write as IOWrite;
+use std::net::TcpStream;
+use std::net::ToSocketAddrs;
+use std::sync::Mutex;
+use std::time::Duration;
+
+type Stream = Box<dyn IOWrite + Sync + Send>;
+
+pub(crate) struct AdvancedTcpStream {
+    hostname: String,
+    port: u16,
+    use_tls: bool,
+    stream: Mutex<Option<Stream>>,
+    connection_timeout: Option<Duration>,
+}
+
+impl AdvancedTcpStream {
+    pub(crate) fn new(
+        hostname: String,
+        port: u16,
+        use_tls: bool,
+        connection_timeout: Option<Duration>,
+    ) -> Self {
+        Self {
+            hostname,
+            port,
+            use_tls,
+            stream: Mutex::new(None),
+            connection_timeout,
+        }
+    }
+
+    pub(crate) fn send_bytes(&self, bytes: &[u8]) -> Result<()> {
+        let mut stream = self.stream.lock().map_err(|err| anyhow!(err.to_string()))?;
+        self.recreate_stream_if_needed(&mut stream)?;
+        if let Some(Err(err)) = stream.as_mut().map(|stream| stream.write_all(bytes)) {
```

Comment:
If the connection was lost, when it will be detected here for the first time. The time elapsed since the actual disconnection can be large, so it's worth to retry the operation. Therefore, I suggest to recreate the TcpStream and repeat the operation once, but only when no recreation occurred on line 37